src/controller/CHIPDeviceController.cpp

```diff
@@ -264,54 +278,61 @@ void DeviceController::OnLocalNOCChainGeneration(void * context, CHIP_ERROR stat
-        ReturnErrorCodeIf(!csrElements.Alloc(kMaxRspLen), CHIP_ERROR_NO_MEMORY);
+    TLV::TLVWriter csrElementWriter;
+    TLV::TLVType containerType;
+    csrElementWriter.Init(csrElements.Get(), kMaxRspLen);
+    ReturnErrorOnFailure(csrElementWriter.StartContainer(TLV::AnonymousTag, TLV::TLVType::kTLVType_Structure, containerType));
+    ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(1), ByteSpan(CSR.Get(), csrLength)));
+
+    // TODO - Need a mechanism to generate CSRNonce for commissioner's CSR
+    ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(2), ByteSpan()));
+    ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(3), ByteSpan()));
+    ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(4), ByteSpan()));
+    ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(5), ByteSpan()));
+    ReturnErrorOnFailure(csrElementWriter.EndContainer(containerType));
+    ReturnErrorOnFailure(csrElementWriter.Finalize());
+
+    mOperationalCredentialsDelegate->SetNodeIdForNextNOCRequest(mLocalDeviceId);
+    mOperationalCredentialsDelegate->SetFabricIdForNextNOCRequest(0);
 
-        TLV::TLVWriter csrElementWriter;
-        TLV::TLVType containerType;
-        csrElementWriter.Init(csrElements.Get(), kMaxRspLen);
-        ReturnErrorOnFailure(csrElementWriter.StartContainer(TLV::AnonymousTag, TLV::TLVType::kTLVType_Structure, containerType));
-        ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(1), ByteSpan(CSR.Get(), csrLength)));
+    // TODO - Need a mechanism to generate signature for commissioner's CSR
+    ReturnErrorOnFailure(
+        mOperationalCredentialsDelegate->GenerateNOCChain(ByteSpan(csrElements.Get(), csrElementWriter.GetLengthWritten()),
+                                                          ByteSpan(), ByteSpan(), ByteSpan(), ByteSpan(), &mLocalNOCChainCallback));
 
-        // TODO - Need a mechanism to generate CSRNonce for commissioner's CSR
-        ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(2), ByteSpan()));
-        ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(3), ByteSpan()));
-        ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(4), ByteSpan()));
-        ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(5), ByteSpan()));
-        ReturnErrorOnFailure(csrElementWriter.EndContainer(containerType));
-        ReturnErrorOnFailure(csrElementWriter.Finalize());
+    ChipLogProgress(Controller, "Called credentials delegate to generate credentials");
+    return CHIP_NO_ERROR;
+}
 
-        mOperationalCredentialsDelegate->SetNodeIdForNextNOCRequest(mLocalDeviceId);
-        mOperationalCredentialsDelegate->SetFabricIdForNextNOCRequest(0);
+CHIP_ERROR DeviceController::LoadLocalCredentials()
+{
+    ChipLogProgress(Controller, "Loading local credentials for fabric %d from the storage", mFabricIndex);
+    ReturnErrorOnFailure(mFabrics.LoadFromStorage(mFabricIndex));
```

Comment:
Is a DeviceController instance associated with a single fabric?

Reply:
Yes, currently it is. There have been discussions on supporting multiple fabrics per controller object. The code needs some refactor to be able to support it.